Барабанова Наталья

NataashaBanan opened the pull request:
@purrrich

NataashaBanan committed:
first realization circle layouter

use spiral

NataashaBanan committed:
Vizualization

Simple vizualization

NataashaBanan committed:
Change spiral in layouter

Remove  square spiral with circle Spiral

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
```
Здесь 2 лишних юзинга

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+            this.Step = step;
+        }
+
+        public IEnumerable<Point> GetNextPoint
```
Решение изящное, но при неаккуратном использовании может уйти в вечный цикл. Лучше `while(true)` в коде держать так, чтобы условие остановки было визуально в пределах видимости. Это нужно для прозрачности и для скорейшего вникания в код другого разработчика.
Плюс название метода этому не помогает, оно наводит на ложную мысль, что метод вернет ровно одну следующую точку на спирали. Контекста класса `Spiral` не хватает, чтобы осознать, а зачем же здесь IEnumerable, да еще и `while(true)`.
Да, сейчас он `internal`, но часто при масштабировании решения возникает необходимость выносить такие штуки в отдельный public-класс.
Как вариант, можно вынести что-нибудь вроде `CurrentAngle` в поля класса, в конструкторе сетить его в 0.0, а в `GetNextPoint` итеративно увеличивать. Ведь в `PutNextRectangle` каждый раз создается новый экземпляр спирали, и нужен он только, чтобы по ней пройтись. Тогда можно будет будет `while(true)` вынести на уровень выше - там, где он по сути и нужен, в `PutNextRectangle`.

P.S. Можем раздуть в личке.

NataashaBanan committed:
Spiral refactoring

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+        public Rectangle PutNextRectangle(Size rectangleSize)
+        {
+            spiral = new Spiral(center, 1);
+            var isSuitiablePoint = true;
```
Здесь можно обойтись без `isSuitiablePoint`. Эта переменная должна вносить понятности в код (и для этих целей названа прекрасно), но конкретно здесь только добавляет лишних строчек кода и развесов, которые визуально мусорят.
У меня получилось переписать этот цикл так, что он похудел в 2 раза.

P.S. Опять же, можем раздуть в личке.

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+            {
+                isSuitiablePoint = true;
+                var newRectangle = new Rectangle(point, rectangleSize);
+                foreach (var rectangle in rectangles)
```
Смотри, вот эту штуку запросто можно вынести из общего цикла в private-метод Layouter-а. Это ведь по сути булочка, которая умеет отвечать на вопрос "Есть ли у нового прямоугольника пересечения с текущими, которые уже уложили?". А все уже уложенные прямоугольники хранятся в Layouter, так что можно выделить отдельный метод с точечной зоной ответственности.
Зачем: это 5 строчек с циклом превратит в одну и читаться будет гораздо лучше.

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+    class CircularCloudLayouter
+    {
+        private Point center;
+        private Spiral spiral;
```
spiral зря здесь выделена полем класса, мы ведь в `PutNextRectangle` каждый раз создаем новую спираль просто чтоб по ней пройтись.
Если хочется обозначить, что мы используем спираль именно с шагом 1, можно шаг вынести в константу.

purrrich commented on cs/TagCloud/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,72 @@
+        [Test]
+        public void PutNextRectangle_ReturnRectangleWithCorrectSize()
+        {
+            var layouter = new CircularCloudLayouter(new Point(0, 0));
```
У всех этих тестов есть один общий кусок, который стоит вынести в `SetUp` - собственно сама тестируемая сущность `CircularCloudLayouter`.
Есть замечательное общепринятое решение: заводить в классе для тестов поле `sut (System Under Test)`. Это позволяет множество тестов в проекте привести к единообразному виду.
![image](https://user-images.githubusercontent.com/73158737/97962677-7f8c3000-1dd7-11eb-99c5-7d041b5a43fb.png)

purrrich commented on cs/TagCloud/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+        
+        [Test]
+        public void PutNextRectangle_ReturnNotIntersectedRectangles()
```
Этот тест является частным случаем следующего.
Давай его выпилим?

purrrich commented on cs/TagCloud/Vizualization.cs:
```diff
@@ -0,0 +1,41 @@
+{
+    public class Vizualization
+    {
+        public static void Main()
```
А зачем здесь Main?..
Проект ведь classLibrary, он не executable.
Если хочется погенерить картинок, можно просто в тесте вызвать Vizualization.Draw() и позапускать его N раз. Коммитить при этом такой тест не надо, только его результаты.
![image](https://user-images.githubusercontent.com/73158737/97963361-c9294a80-1dd8-11eb-82ef-37401b43cb37.png)

purrrich commented on cs/TagCloud/Vizualization.cs:
```diff
@@ -0,0 +1,41 @@
+                height = (int)Math.Round(height/1.1);
+            }
+
+            bitmap.Save("C:\\Users\\Наталья\\Desktop\\myPng.bmp");
```
1. Для экранирования слэшей можно использовать волшебную собаку `@"C:\Users\Наталья\Desktop\myPng.bmp"`.
2. Давай сделаем этот путь относительным (пусть лежит в директории проекта) и запретим перетирать старый файл? Можно сделать название файла зависимым от даты и времени, например. Или генерить им уникальные названия каким-то другим способом.

purrrich commented on cs/TagCloud/Vizualization.cs:
```diff
@@ -0,0 +1,41 @@
+            bitmap.Save("C:\\Users\\Наталья\\Desktop\\myPng.bmp");
+        }
+
+        public static void DrawAndFillRectangle(Graphics graphics, Rectangle rectangle)
```
А он точно не private?..

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+    {
+        private Point center;
+        private Spiral spiral;
+        private List<Rectangle> rectangles = new List<Rectangle>();
```
Давай `new List<Rectangle>()`, а все поля сделаем readonly?

purrrich commented on cs/TagCloud/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+{
+    internal class Spiral
+    {
+        private Point Center;
```
1. Здесь поля класса в большой буквы (`CamelCase`), а в `CircularCloudLayouter` - с маленькой (`camelCase`). Надо выбрать что-то одно)) У разных команд кодстайл разный, так что какой именно вариант решать тебе, главное чтобы было единообразно.
Ну и да, если это будет `CamelCase`, то `this.` в конструкторе лишнее.
2. Их можно сделать `readonly`.

purrrich commented on cs/TagCloud/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,72 @@
+        [Test]
+        public void PutNextRectangles_LayoutRectanglesInCircle()
+        {
+            var center = new Point(0, 0);
```
Не хочется ли здесь проверить приближенность прямоугольников к какой-нибудь другой точке, более интересной чем (0, 0)?

purrrich commented on cs/TagCloud/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,72 @@
+        public void PutNextRectangle_ManyRectangles_ReturnNotIntersectRectangles()
+        {
+            var layouter  = new CircularCloudLayouter(new Point(0, 0));
+            var size = new Size(2, 1);
```
Здесь можно не привязываться к конкретным размерам, пусть тест генерит их рандомно.
В таком случае, если для каких-то других входных данных алгоритм работает неверно, тест хотя бы начнет плавать.

NataashaBanan committed:
Move while(true)

from Spiral to putNextRectangle

NataashaBanan committed:
Edit Drawing path

1. Relative path
2. Generate png name using milliseconds

NataashaBanan committed:
Refactor

1. Delete method Main from vizualization
2. Delete all constants from vizualizator
3. Refactor fields in Layouter

NataashaBanan committed:
Add README

Add readme and test vizualization

NataashaBanan committed:
AddingTearDown

Also change draw method in  vizualizator

NataashaBanan committed:
fix mistaces

Add abs to color counter and remove redundant argument from Draw.
